### ISSUE TITLE
Accessibility service: this service is malfunctioning

### ISSUE THREAD

**Issue by actuallymentor** [bitwarden/android#596]
My accessibility service (android 9) keeps breaking. When I go to the `enable` screen android notifies me that `this service is malfunctioning`.

Let me know if I can supply any useful debug data.

**Comment by kspearrin** [comment]
What type of device?

**Comment by actuallymentor** [comment]
Oneplus 3, Android 9 (stock).

**Comment by htfy96** [comment]
Also appears on my Zenfone 6 with Android 9. There's no icon at status bar in the presence of password box. Reabling Bitwarden in accessibility can fix this problem for a while, but the problem reappears after several minutes.

**Comment by actuallymentor** [comment]
> Reabling Bitwarden in accessibility can fix this problem for a while, but the problem reappears after several minutes.

Confirming that this is the same for me. I can re-enable but after a time (unpredictable) it breaks again.

**Comment by vargbeaumont** [comment]
@actuallymentor if you're using a OnePlus device go into the battery settings, battery optimisation, click on the three dots in the top right, click advanced optimisation then disable deep optimisation. This seems to have fixed this issue for me.

**Comment by actuallymentor** [comment]
@vargbeaumont thanks for the tip!

> if you're using a OnePlus

I am.

> disable deep optimisation

Did it just now and will post back if it ends up breaking.

**Comment by idavydov** [comment]
I have the same problem. I tried to keep deep optimization, but select "don't optimize" next to bitwarden optimization settings (settings -> battery -> battery optimisation). Will report back if I still encounter the problem.

**Comment by actuallymentor** [comment]
> Did it just now and will post back if it ends up breaking.

Confirming that even unoptimised it keeps crashing.

No other accessibility services have this issue currently.

**Comment by SOLR4189** [comment]
Did somebody find a solution for this problem? I have the same problem

**Comment by Web-E** [comment]
Same issue here as well, OnePlus 3 (android 9) + battery optimization turned off

**Comment by AtjonTV** [comment]
Same issue on HUAWEI P smart Z (STK-LX1) with EMUI 9.1 (Android 9).

**Comment by onfire4g05** [comment]
Also can confirm this is happening on my Google Pixel 3 XL.

**Comment by MikeBishop** [comment]
OnePlus 8 running Android 11 checking in, same issue.

**Comment by quamok** [comment]
Pixel 4 xl, android 11, same issue here also.

**Comment by Ankurjain30** [comment]
Hi Guys 
I also starting developing once and just trying to get kind of event and getting service is malfunctioning on oneplus 8T. Did u able to resolve the issue somehow.

**Comment by KaiLCU** [comment]
Identical problem on Poco F3 running MIUI 12.5 (Android 12)

**Comment by rohitkhullar88** [comment]
Facing same issue in sending WhatsApp message using accessibility service but it crashes after 5 times working properly. Then I need to manually click 2 times back button to come back to my Original Android App after sending the message.

**Comment by maxgrass** [comment]
Redmi Note 9
Same message while trying to click the icon in quick settings.

**Comment by AM-ElZaher** [comment]
Same error on Xiaomi Mi 9T
Is there any fix soon?

**Comment by SudoNova** [comment]
Same issue MiUI12.5.5 on Mi 10T (Android 11)

**Comment by gnongsiej** [comment]
Not sure if this will help anyone but you need to check the accessibility shortcut and set it to accessibility button.

**Comment by nri82** [comment]
Same here MIUI 12.5 android 11, Poco X3 NFC

Clear app data, reinstall, reboot did not solves this problem.

![IMG_20220118_192711](https://user-images.githubusercontent.com/18355195/149978814-61bd8782-e6ee-4e5d-82f3-f845df1c160f.jpg)

**Comment by HydroFerret** [comment]
No go on Poco X3 Pro. Whats the deal yo?

**Comment by EMREOYUN** [comment]
The same happens on Huawei P Smart 2018 with EMUI 9.1 (Android 9)

**Comment by nri82** [comment]
Okay I found a simple solution.

Just give auto start permission and then problem solved.

![Screenshot_20220630_073231](https://user-images.githubusercontent.com/18355195/176593376-bb4cfc2c-c000-48e3-bf45-9989b1c6689b.jpg)

**Comment by blakeNaccarato** [comment]
@nri82 What phone and OS version are you running? I don't see that exact interface on my Pixel 3a running Android 12 (Build SP2A.220505.006). I wonder if there's a similar permission that goes by a different name?

**Comment by nri82** [comment]
> @nri82 What phone and OS version are you running? I don't see that exact interface on my Pixel 3a running Android 12 (Build SP2A.220505.006). I wonder if there's a similar permission that goes by a different name?

Poco X3 NFC, android 11, MIUI 12.5. in short you need to give auto start permission but I don't know it's works or not on your situation.

**Comment by EMREOYUN** [comment]
> Okay I found a simple solution.
> 
> Just give auto start permission and then problem solved.

Autostart is not a thing on Huawei

**Comment by nri82** [comment]
> > Okay I found a simple solution.
> > Just give auto start permission and then problem solved.
> 
> @nri82 Autostart is not a thing on Huawei

Yes [I noticed that](https://dontkillmyapp.com/huawei) my solution is working on Xiaomi MIUI sorry for this situation.

**Comment by Web-E** [comment]
How about using some Automator like "Tasker" to autostart the the app, say at interval of 1 hour. That might work.
I no longer face this issue as I moved form OP3 to OP8

**Comment by EMREOYUN** [comment]
> How about using some Automator like "Tasker" to autostart the the app, say at interval of 1 hour. That might work.
> I no longer face this issue as I moved form OP3 to OP8

@Web-E Huawei also does not like these "Automators" and eventually kill those as well.

**Comment by David263** [comment]
Using Tasker to solve this problem probably won't work. Tasker has the same problem.

**Comment by sameert89** [comment]
Android 12 MIUI, same problem. Not just this app some other apps' accessibility is broken.

**Comment by Turab** [comment]
1- Force stop the application
2- Enable auto start
3- Enable battery saver first, then disable again.
4- Now enable accessibility again.

This solved the problem for MIUI 13 Android 12. The reason for this malfunction is;

1- Google is trying to cripple call recording apps and constantly limiting the use of Accessibility API
2- Some of the vendors like Xiaomi and Samsung kill background applications too often to "save" battery and this breaks how the app works.

By the way, although everything is working smoothly, Bitwarden fails to autofill password fields which open only numpad for entry.

**Comment by ameerhamza6733** [comment]
> Okay I found a simple solution.
> 
> Just give auto start permission and then problem solved.
> 
> ![Screenshot_20220630_073231](https://user-images.githubusercontent.com/18355195/176593376-bb4cfc2c-c000-48e3-bf45-9989b1c6689b.jpg)

Yes, its work

**Comment by David263** [comment]
Sometimes Accessibility service fails for me, but when using Tasker, not Bitwarden. Also, I can't find this AutoStart option in my app info screen, or in the Permissions Manager. Android 11, running on a TCL A3. The Tasker people say it's due to a complicated bug in Android, best as I understand them.

**Comment by jf8989** [comment]
> 

Hi there.  Thank you for clarifying what the issue is.  Regarding any accessibility service, it's just impossible to re-enable them automatically if they either malfunction and/or are turned off, isn't it?  At least without user interaction, I believe it's impossible to do so, am I right?  Now, when you say "enable and disable battery saver", do you mean we ought to do that with the global battery saver or just within the app's settings?  I've been experiencing the same issue  on my Xiaomi 13 / Android 12 (RemiNote9).  But I do wonder whether this malfunction would also happen on Android 13+.  Androi 9 seems to be fine, but I yet have to test Android 10, and 11.

**Comment by David263** [comment]
Unfortunately, I'm not an expert on how the Android accessibility service is programmed, or any of Android, for that matter. I'm just a user of Tasker and Bit Warden (and a retired professional software engineer) and not happy with the mixed uses of the accessibility service. Android wants accessibility service to be just for disabled people, but there are certain apps, like Tasker and Bit Warden, that want to be helpful by making devices more useful for everyone. Ideally, Android should provide an "internal" or "advanced" service for such Apps, separate from accessibility service. We need one or two more Permission categories, probably. But only Android developers can help, and they seem unwilling to. Developers of apps like Tasker and Bit Warden must kludge together solutions to get around current Android limitations. The situation is unfair, but not an unusual one in software development history.

**Comment by David263** [comment]
It seems to me that one such kludge is to run a separate service or app that monitors the accessibility service and re-enables it when Android decides to stop it or to block permission to use it. Such a service only needs to run once every few seconds.

**Comment by jf8989** [comment]
> It seems to me that one such kludge is to run a separate service or app that monitors the accessibility service and re-enables it when Android decides to stop it or to block permission to use it. Such a service only needs to run once every few seconds.

Well, as far as I'm concerned, it's impossible to re-enable an accessibility service without user interaction after it has been turned off.  Nevertheless, I don't know if it's possible to automatically fix it if the problem is that it is malfunctioning while still being on.  In this case, theoretically, one should be able to do something in order to automatically repair the malfunction.  Maybe a work manager could fix it.  I'm sure it's possible to monitor the service state at the least.   I can't believe no one has found a workaround.  Or maybe they have and haven't shared it yet.

**Comment by jf8989** [comment]
Now that I think about it, it'd make sense to be cautious not to share the workaround because android developers would just patch it.  I mean, clearly they have no intention of prioritizing continuity on Android services in general.

**Comment by David263** [comment]
Over at least the past year, the Tasker developer has frequently posted supposed fixes for the problem that restarting Android or installing a new version sometimes stops the accessibility service, including his creating a separate app that the user installs. None of these fixes have worked perfectly for me, although some have helped. I agree that it may not be ethical to publicize a fix, since it currently seems against the wishes of the Android developers. As I said, there needs to be a new and separate service for apps like Tasker and Bit Warden to give them the ability to do what they need to do.

**Comment by jf8989** [comment]
> Over at least the past year, the Tasker developer has frequently posted supposed fixes for the problem that restarting Android or installing a new version sometimes stops the accessibility service, including his creating a separate app that the user installs. None of these fixes have worked perfectly for me, although some have helped. I agree that it may not be ethical to publicize a fix, since it currently seems against the wishes of the Android developers. As I said, there needs to be a new and separate service for apps like Tasker and Bit Warden to give them the ability to do what they need to do.

I totally agree with you.  Not only would it be unethical, it'd also be a problem to the one that found the fix because Android would take a look and patch it.  If I found a fix, I wouldn't want to let Android know.  I mean, I understand their posture on this matter; nevertheless, like you well said, they should at least allow some apps to use an unrestricted type of service that let's them do what they are intending to do.  I agree 100%.  If anyone would like to create an entire new mobile OS, please let me know so we can team up because I'm tired of Android...he he.

**Comment by andymarden** [comment]
Had this issue myself with poco F3, miui 14 and bitwarden.

Autostart (once I be found the setting which moved) didn't fix it, Bitwarden still shows it is  malfunctioning.

Tried to lock the app but there is this bug on the recent apps and 3rd party launchers in the menu when you long press dues but sure.

Eventually found it (miui seen to have a competition at each release to move a setting to the most obscure place possible).

That works - is you lock Bitwarden in memory, the accessibility is fixed. 

Why that should error out is this is but dry must, in itself, be a bug.

Locking the app (not too be confused with lock app which is completely different ffs) can now be found under Security -> Boost Speed -> Lock Apps.

Can I just say how much I enjoy, in this screen, the fact that the apps are listed in NI discernable order. And just wtf has Boost Speed got to do with Security?

It's as if they are trying to make it impossible to find and, if you are lucky enough to stumble across it, unnecessarily awkward to use. Jesus, Xiaomi..

Still - that's how I got it working.
